trade.lic

```diff
@@ -205,58 +697,852 @@ class Trade
+    end
+    item_container = @settings.storage_containers.find { |container| bput("tap my #{@crafted_item.split.last} in my #{container}", 'unfinished', 'You tap', 'I could not find') == 'unfinished' }
+    if item_container
+      case bput("get my #{@crafted_item.split.last} from my #{item_container}", 'You get', 'What were', 'But that')
+      when 'You get'
+        echo "Engineering found an uncompleted item #{@crafted_item}"
+        be_inside_caravan
+        magic_cleanup
+        start_script('shape', ['continue', @crafted_item.split.last])
+        return
+      end
+    end
+
+    recipes = get_data('recipes').crafting_recipes.select { |recipe| recipe['type'] =~ /shaping/i }
+    recipe = recipe_lookup(recipes, @crafted_item)
+    if recipe['volume'] > @current_lumber_count
+      wipe_skill('Engineering', @lead_training_skills)
+      return engineering_cleanup
+    else
+      be_inside_caravan
+      magic_cleanup
+      @current_lumber_count -= recipe['volume']
+      start_script('shape', ['trash', recipe['chapter'], recipe['name'], 'balsa', recipe['noun']].map { |arg| arg =~ /\s/ ? "\"#{arg}\"" : arg })
+    end
+  end
+
+  def engineering_cleanup # sanitizes engineering
+    return unless @training_token == 'Engineering'
+    stop_script('shape') if Script.running?('shape')
+    bput("drop my #{@crafted_item.split.last}", 'You drop', 'You place', 'What were', 'You spread') unless bput("tap my #{@crafted_item.split.last}", 'unfinished', 'You tap', 'I could not find') == 'unfinished'
+    stow_item(right_hand) if right_hand
+    stow_item(left_hand) if left_hand
+    stow_hands
+    bput('get my balsa lumber', 'You get', 'You pick up', 'You are already', 'What were you')
+    stow_hands
+    be_outside_caravan
+    wipe_token
+    magic_cleanup
+    @times_engineering += 1
+  end
+
+  def first_aid_routine # runs ;first-aid on compendiums or textbooks
+    return unless @training_token == 'First Aid'
+    return first_aid_cleanup if DRSkill.getxp('First Aid') > 30
+    return first_aid_cleanup if @cooldown_timers['fa-timer'] && !Script.running?('first-aid')
+    @cooldown_timers['fa-timer'] ||= Time.now
+    return if Script.running?('first-aid')
+    be_inside_caravan
+    start_script('first-aid')
+  end
+
+  def first_aid_cleanup
+    return unless @training_token == 'First Aid'
+    echo "first_aid cleanup" if debugging?
+    wipe_token
+    stop_script('first-aid') if Script.running?('first-aid')
+    @cooldown_timers['First Aid'] = Time.now
+    @cooldown_timers['fa-timer'] = nil
+    waitrt?
+  end
+
+  def performance_routine # runs ;performance for 2 minutes.  Will run ;first-aid too, if it hasn't been done very recently.
+    return unless @training_token == 'Performance'
+    echo "performance routine" if debugging?
+    if @caravan_being_led && check_ability?('First Aid', 180)
+      @training_token = 'First Aid' if @settings.textbook ? exists?(@settings.textbook_type) : exists?('compendium')
```

Comment:
If I dont have either, this results in spam tapping:

```
 [trade]>tap my compendium                                                          
 I could not find what you were referring to.                                       
 >                                                                                  
 [trade: in lead training - training token is Performance]                          
 [trade: performance routine]                                                       
 [trade]>tap my compendium                                                          
 I could not find what you were referring to.                                       
 >                                                                                  
 [trade: in lead training - training token is Performance]                          
 [trade: performance routine]                                                       
 [trade]>tap my compendium                                                          
 I could not find what you were referring to.                                       
```

Reply:
Yes, thanks for finding that.  I only wanted to put something in base if it was 1) likely for a young trader to have, and 2) not going to cause a script failure if they didn't have it.  But it's so easy to throw a check for compendiums or textbooks in startup validation that I think I'll just do that.

Reply:
Among a couple other little bugs, there's a check for if you have a compendium/textbook on startup, and if you don't it'll wipe FA from your training list and not attempt to ever try it.